exercises/matching-brackets/canonical-data.json

```diff
@@ -118,6 +118,19 @@
       },
       "expected": false
     },
+    {
+      "uuid": "1d5c093f-fc84-41fb-8c2a-e052f9581602",
+      "reimplements": "a0205e34-c2ac-49e6-a88a-899508d7d68e",
```

Comment:
The new test is fine, but I don't see why it needs to reimplement another one. Why not have both?

Reply:
I guess my reasoning when chosing to reimplement the test was that having only the new one better conveyed the intent behind the problem, I thought that maybe the first implementating could lure the user to code the wrong answer - or something like that.

That being said, maybe having this redundancy is a better way of asserting a valid implementation and might even help the user to understand the problem in a more didactic way (step by step) - given the user went with a wrong implementation.

Honestly, I didn't give it much thought because it seemed to me that's exactly the kind of dilema better discussed with the people that maintain exercism (you guys). Maybe I should've shared this toughts when creating the PR, instead of just raising the possibility of making a new test for the issue.

I would happily change it to implement a new test if you guys think that's a better approach to it.

Reply:
I don't mind the old test being reimplemented because I didn't see `[({]})` (old test) testing anything that `[({}])` (new test) doesn't already cover. If any other test is warranted in addition to this new test, it would actually be one where the outermost ones are correct, `([{]})`. The implementation that that test is designed for is even more "creatively wrong" though...

```ruby
  {
    # https://github.com/exercism/problem-specifications/pull/2071
    name: 'count only, but also checks last opened',
    should_fail: true,
    f: ->(i, _) {
      open = BracketMatcher::PAIRS.transform_values { 0 } 
      last_opened = nil 
      i['value'].each_char { |c| 
        if open.has_key?(c)
          last_opened = c 
          open[c] += 1
        end 
        if (o = BracketMatcher::CLOSES[c])
          return false if open[o] == 0
          return false if last_opened &.!= o
          last_opened = nil 
          open[o] -= 1
        end 
      }   
      open.values.all?(&:zero?)
    },  
  },  
  {
    # https://github.com/exercism/problem-specifications/pull/2071
    name: 'count only, but also checks first opened',
    should_fail: true,
    f: ->(i, _) {
      open = BracketMatcher::PAIRS.transform_values { 0 } 
      first_opened = nil 
      i['value'].each_char { |c| 
        if open.has_key?(c)
          first_opened ||= c
          open[c] += 1
        end 
        if (o = BracketMatcher::CLOSES[c])
          return false if open[o] == 0
          open[o] -= 1
          if open.values.all?(&:zero?)
            return false if first_opened != BracketMatcher::CLOSES[c]
            first_opened = nil 
          end 
        end 
      }   
      open.values.all?(&:zero?)
    },  
  }, 
```

Reply:
> I don't mind the old test being reimplemented because I didn't see `[({]})` (old test) testing anything that `[({}])` (new test) doesn't already cover. If any other test is warranted in addition to this new test, it would actually be one where the outermost ones are correct, `([{]})`. The implementation that that test is designed for is even more "creatively wrong" though...
> 
> ```ruby
>   {
>     # https://github.com/exercism/problem-specifications/pull/2071
>     name: 'count only, but also checks last opened',
>     should_fail: true,
>     f: ->(i, _) {
>       open = BracketMatcher::PAIRS.transform_values { 0 } 
>       last_opened = nil 
>       i['value'].each_char { |c| 
>         if open.has_key?(c)
>           last_opened = c 
>           open[c] += 1
>         end 
>         if (o = BracketMatcher::CLOSES[c])
>           return false if open[o] == 0
>           return false if last_opened &.!= o
>           last_opened = nil 
>           open[o] -= 1
>         end 
>       }   
>       open.values.all?(&:zero?)
>     },  
>   },  
>   {
>     # https://github.com/exercism/problem-specifications/pull/2071
>     name: 'count only, but also checks first opened',
>     should_fail: true,
>     f: ->(i, _) {
>       open = BracketMatcher::PAIRS.transform_values { 0 } 
>       first_opened = nil 
>       i['value'].each_char { |c| 
>         if open.has_key?(c)
>           first_opened ||= c
>           open[c] += 1
>         end 
>         if (o = BracketMatcher::CLOSES[c])
>           return false if open[o] == 0
>           open[o] -= 1
>           if open.values.all?(&:zero?)
>             return false if first_opened != BracketMatcher::CLOSES[c]
>             first_opened = nil 
>           end 
>         end 
>       }   
>       open.values.all?(&:zero?)
>     },  
>   }, 
> ```

Both cases (and also the original intent of the test) are about phrases where even though all brackets are indeed closed, they are closed in the wrong order. Should we change my reimplementation to solve both cases (with something like `([({}]))` - I'm surrounding the proposed idea with a correct set of outermost brackets), or you would prefer new tests being added? 

One approach would be to (1) change my reimplementation of "paired and wrong nested" to be a new test like "paired and deep wrong nested" (`[({}])`) and (2) then to add another test for "paired and wrong nest within outermost right" (`([{]})`). If we change the reimplementation to solve both edge cases, it might start to get confusing on why the test input is the way it is. It might be better for the user (which is learning) if they were actually separated tests for specific problems.

That being said though, IMO, I would still say to reimplement the one test ("paired and wrong nested") for all edge cases, not only they all refer to the same idea, I tend to believe it's not that big of a deal for the user to realize what they did wrong given they went with a wrong solution. Not only that, but these wrong solutions have to be so creative that most people would probably not understand - and maybe would get confused - on why there is a test for "deep nested" and "wrong within outermost right brackets" - and I don't think they should, these tests are not for them, they are for those who went with these wrong solutions. For explicability, if anyone ever asks, there would be two comments in the canonical JSON, one for each of the edge cases. That's why comments are there for.

What do you think? What do you do in those cases? Reimplements or create new ones?

Reply:
My guess is that it's better for each case to only test one thing because let's say a student submits an implementation that's incorrect in only one way. The student gets a better indication of what went wrong if the test case that failed is focused instead of trying to cover multiple incorrect implementations.

However, it is only a guess. I don't think we've ever had data to answer the question "do students get confused by test cases that are not meant for them, but only for particularly creative wrong solutions?"

If there is doubt about the value of `([{]})`, it doesn't necessarily need to be added until there is an actual sighting of a submitted solution that would get it wrong.

Reply:
Having a better indication of what went wrong is indeed better. I agree.